dimod/generators/satisfiability.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2022 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from __future__ import annotations
+
+import collections.abc
+import itertools
+import math
+import typing
+
+import numpy as np
+
+import dimod  # for typing
+
+from dimod.binary import BinaryQuadraticModel
+from dimod.vartypes import Vartype
+
+
+__all__ = ["nae3sat"]
+
+
+def _iter_interactions(num_variables: int, num_clauses: int,
+                       seed: typing.Union[None, int, np.random.Generator] = None,
+                       ) -> typing.Iterator[typing.Tuple[int, int, int]]:
+
+    rng = np.random.default_rng(seed)
+
+    for _ in range(num_clauses):
+        x, y, z = rng.choice(num_variables, 3, replace=False)
+        xs, ys, zs = 2 * rng.integers(0, 1, endpoint=True, size=3) - 1
+        yield (x, y, xs*ys)
+        yield (x, z, xs*zs)
+        yield (y, z, ys*zs)
+
+
+def _iter_interactions_without_replacement(
+        num_variables: int, num_clauses: int,
+        seed: typing.Union[None, int, np.random.Generator] = None,
+        ) -> typing.Iterator[typing.Tuple[int, int, int]]:
+
+    try:
+        max_num_clauses = math.comb(num_variables, 3)
+    except AttributeError:
+        # Python < 3.8
+        # This is a bit less picky and has overflow issues but
+        # probably fine for any cases users are likely to actually
+        # submit
+        max_num_clauses = math.factorial(num_variables)
+        max_num_clauses //= (6 * math.factorial(num_variables - 3))
+
+    if num_clauses > max_num_clauses:
+        raise ValueError("rho results in too many clauses")
+
+    rng = np.random.default_rng(seed)
+
+    seen: typing.Set[typing.FrozenSet[int]] = set()
+    while len(seen) < num_clauses:
+        x, y, z = rng.choice(num_variables, 3, replace=False)
+
+        fz = frozenset((x, y, z))  # so we're order-invarient
+        if fz in seen:
+            continue
+        seen.add(fz)
+
+        xs, ys, zs = 2 * rng.integers(0, 1, endpoint=True, size=3) - 1
+
+        yield (x, y, xs*ys)
+        yield (x, z, xs*zs)
+        yield (y, z, ys*zs)
+
+
+def nae3sat(variables: typing.Union[int, typing.Sequence[dimod.typing.Variable]],
```

Comment:
I suggested using rho instead of num_clauses since that is the indicator of the problem class. Otherwise the user would need to add boilerplate when generating problems. However, @jackraymond  is the expert on this problem class, so I ultimately defer to him.

Reply:
fwiw, we went with deterministic num_clauses from rho (rounding num_variables*rho) instead of the random distribition approach

Reply:
Personally, I think its easier and more intuitive to define the class in terms of a fixed integer number of clauses, rather than a random number of clauses (determined from distribution parameterized by rho - which is buried deep in code and not clearly documented).

Reply:
yes, we'll just add using rho as an example in docs